eventstore_test.go

```diff
@@ -106,76 +105,90 @@ func TestEventStoreWithRegistry(t *testing.T) {
 				data, ok := events[0].Data.(*OrderPlaced)
 				is.True(ok)
 				is.Equal(*data, devent)
 			},
 		},
 		{
-			"append-load-with-occ",
+			"append-expect-sequence",
 			func(t *testing.T, es *EventStore) {
 				ctx := context.Background()
 
-				event1 := &Event{
-					Entity: "order.123",
-					Data:   &OrderPlaced{ID: "123"},
-					Meta: map[string]string{
-						"geo": "eu",
-					},
-				}
-
-				seq, err := es.Append(ctx, []*Event{event1}, ExpectSequence(0))
+				seq, err := es.Append(ctx, []*Event{{
+					Entity: "order.1",
+					Data:   &OrderPlaced{},
+				}}, ExpectSequence(0))
 				is.NoErr(err)
 				is.Equal(seq, uint64(1))
 
-				event2 := &Event{
-					Entity: "order.123",
-					Data:   &OrderShipped{ID: "123"},
-				}
-
-				seq, err = es.Append(ctx, []*Event{event2}, ExpectSequence(1))
+				seq, err = es.Append(ctx, []*Event{{
+					Entity: "order.1",
+					Data:   &OrderShipped{},
+				}}, ExpectSequence(1))
 				is.NoErr(err)
 				is.Equal(seq, uint64(2))
-
-				var events eventSlice
-				lseq, err := es.Evolve(ctx, "order", &events)
-				is.NoErr(err)
-
-				is.Equal(seq, lseq)
-				is.Equal(len(events), 2)
 			},
 		},
 		{
-			"append-load-partial",
+			"append-expect-sequence-subject",
 			func(t *testing.T, es *EventStore) {
 				ctx := context.Background()
 
-				seq, err := es.Append(ctx, []*Event{{Entity: "order.123", Data: &OrderPlaced{ID: "123"}}})
+				seq, err := es.Append(ctx, []*Event{{
+					Entity: "order.1",
+					Data:   &OrderPlaced{},
+				}})
 				is.NoErr(err)
 				is.Equal(seq, uint64(1))
 
-				seq, err = es.Append(ctx, []*Event{{Entity: "order.123", Data: &OrderShipped{ID: "123"}}})
+				seq, err = es.Append(ctx, []*Event{{
+					Entity: "order.1",
+					Data:   &OrderShipped{},
+				}}, ExpectSequence(1)) // specific entity
 				is.NoErr(err)
 				is.Equal(seq, uint64(2))
 
-				var events eventSlice
-				lseq, err := es.Evolve(ctx, "order", &events, AfterSequence(1))
+				seq, err = es.Append(ctx, []*Event{{
+					Entity: "order.2",
+					Data:   &OrderPlaced{},
+				}}, ExpectSequenceSubject(2, "order")) // for all orders
 				is.NoErr(err)
+				is.Equal(seq, uint64(3))
 
-				is.Equal(seq, lseq)
-				is.Equal(len(events), 1)
-				is.Equal(events[0].Type, "order-shipped")
+				seq, err = es.Append(ctx, []*Event{{
+					Entity: "order.3",
+					Data:   &OrderPlaced{},
+				}}, ExpectSequenceSubject(0, "order.3")) // relative to event entity (default)
+				is.NoErr(err)
+				is.Equal(seq, uint64(4))
+
+				seq, err = es.Append(ctx, []*Event{{
+					Entity: "order.4",
+					Data:   &OrderPlaced{},
+				}}, ExpectSequenceSubject(4, "order.3")) // relative to a different entity
+				is.NoErr(err)
+				is.Equal(seq, uint64(5))
+
+				seq, err = es.Append(ctx, []*Event{{
+					Entity: "order.5",
+					Data:   &OrderPlaced{},
+				}}, ExpectSequenceSubject(2, "*.*.order-shipped")) // relative to a type
+				is.NoErr(err)
+				is.Equal(seq, uint64(6))
 			},
 		},
+		/* TODO
```

Comment:
Deduping with atomic batch publish has a different behavior than normal publish dedupe, so I need to revisit this.